src/libos/src/events/waiter_queue.rs

```diff
@@ -0,0 +1,66 @@
+use std::collections::VecDeque;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::time::Duration;
+
+use super::{Waiter, Waker};
+use crate::prelude::*;
+
+/// A queue for waiters.
+///
+/// By using this queue, we can wake up threads in their waiters' enqueue order.
+pub struct WaiterQueue {
+    count: AtomicUsize,
+    wakers: SgxMutex<VecDeque<Waker>>,
```

Comment:
I would like keep the name `WaiterQueue`. It is better in my opinion. The name emphasizes that it is `Waiter`s that are enqueued and dequeued.

Reply:
Thanks for the above explanations. I started to be lost from here. Using `Waiter` here can also implement the current functions. `Waiter` is powerful. It can also get `Waker` but not verse vice.

Reply:
`Waiter` implements `!Send` and `!Sync` traits. So it cannot be used as items in `WaiterQueue`. As to why `!Send` and `!Sync` traits are implemented for `Waiter`, see my response above.